L82-core-event-engine.md

```diff
@@ -0,0 +1,538 @@
+gRPC Core EventEngine API
+----
+* Author: AJ Heller (@drfloob)
+* Approver: Mark Roth (@markdroth)
+* Status: In Review
+* Implemented in: gRPC Core
+* Last updated: 2021.06.15
+* Discussion at: https://groups.google.com/g/grpc-io/c/EamJ_ae_db0
+
+## Abstract
+
+This work replaces gRPC Core's iomgr with a public interface for custom,
+pluggable implementations which we're calling EventEngines. EventEngines are
+tasked with providing all cross-platform I/O, task execution, and DNS resolution
+functionality for gRPC Core and its wrapped languages.
+
+Unlike the current polling model where threads are borrowed from the
+application, EventEngines will bring their own polling and callback execution
+threads. This public API will make it easier to integrate gRPC into external
+event loops, it will support siloing events between gRPC channels and servers,
+and it will provide another way to support the C++ Callback API.
+
+## Background
+
+gRPC Core and its wrapped languages rely internally on an event management
+framework called iomgr. A small collection of homegrown, platform-specific
+implementations exist, allowing I/O functionality to be customized to the
+various platforms that gRPC supports (IOCP is used on Windows, epoll on Linux,
+poll on Mac, etc).
+
+Over the past few years, there have been multiple projects aimed at improving or
+replacing iomgr. The EventEngine effort takes a slightly different approach to
+things, but otherwise shares the same rationale as the previous effort.
+
+### Related Proposals:
+
+* This work supersedes [L69: EventManger API](https://github.com/grpc/proposal/pull/182)
+* Supports [L67: C++ callback-based asynchronous API](https://github.com/grpc/proposal/pull/180) as a more performant solution than provided by the [CallbackAlternativeCQ](https://github.com/grpc/grpc/pull/25169).
+
+## Proposal
+
+### Concepts
+
+An EventEngine is roughly composed of:
+
+* A task runner which asynchronously executes callables with optional
+  scheduling.
+* An asynchronous DNS resolver.
+* Network listener logic that accepts incoming connection requests, and
+  performs network-server-related I/O operations.
+* Client logic, support for making outgoing network connections.
+
+### Common Data Types
+
+```cpp
+namespace grpc_event_engine {
+namespace experimental {
+
+/// A set of parameters used to configure an endpoint, either when initiating a
+/// new connection on the client side or when listening for incoming connections
+/// on the server side. An EndpointConfig contains a set of zero or more
+/// Settings. Each setting has a unique name, which can be used to fetch that
+/// Setting via the Get() method. Each Setting has a value, which can be an
+/// integer, string, or void pointer. Each EE impl should define the set of
+/// Settings that it supports being passed into it, along with the corresponding
+/// type.
+class EndpointConfig {
```

Comment:
They all share the same `grpc_event_engine::experimental` namespace, which provides logical grouping. I don't know of any good reasons we'd want to weaken encapsulation by allowing privileged access from inner classes to EventEngine internals (should it become an abstract class instead of pure interface). There's also the added verbosity of the resulting code if the classes are nested (an extra `EventEngine::` peppered around).

Do you have arguments as to why we should nest all related classes within the EventEngine class?